### ISSUE TITLE
vroom::spec returns NULL when col_select is specified and readr is loaded

### ISSUE THREAD

**Issue by kent37** [r-lib/vroom#303]
When `readr` is loaded, calling `vroom::vroom` with a `col_select` argument returns a data frame without a `spec` attribute so `vroom::spec()` returns `NULL`.

``` r
library(vroom)

# This works
df <- vroom(vroom_example("mtcars.csv"), col_select=-mpg)
#> Rows: 32
#> Columns: 11
#> Delimiter: ","
#> chr [ 1]: model
#> dbl [10]: cyl, disp, hp, drat, wt, qsec, vs, am, gear, carb
#> 
#> Use `spec()` to retrieve the guessed column specification
#> Pass a specification to the `col_types` argument to quiet this message

vroom::spec(df)
#> cols(
#>   model = col_character(),
#>   mpg = col_skip(),
#>   cyl = col_double(),
#>   disp = col_double(),
#>   hp = col_double(),
#>   drat = col_double(),
#>   wt = col_double(),
#>   qsec = col_double(),
#>   vs = col_double(),
#>   am = col_double(),
#>   gear = col_double(),
#>   carb = col_double(),
#>   .delim = ","
#> )

library(readr)
#> Registered S3 methods overwritten by 'readr':
#>   method           from 
#>   format.col_spec  vroom
#>   print.col_spec   vroom
#>   print.collector  vroom
#>   print.date_names vroom
#>   print.locale     vroom
#>   str.col_spec     vroom
#> 
#> Attaching package: 'readr'
#> The following objects are masked from 'package:vroom':
#> 
#>     as.col_spec, col_character, col_date, col_datetime, col_double,
#>     col_factor, col_guess, col_integer, col_logical, col_number,
#>     col_skip, col_time, cols, cols_condense, cols_only, date_names,
#>     date_names_lang, date_names_langs, default_locale, fwf_cols,
#>     fwf_empty, fwf_positions, fwf_widths, locale, output_column,
#>     problems, spec

# This returns a data frame with no 'spec' attribute
df <- vroom(vroom_example("mtcars.csv"), col_select=-mpg)

vroom::spec(df)
#> NULL
```

<sup>Created on 2021-02-03 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> - Session info ---------------------------------------------------------------
#>  setting  value                       
#>  version  R version 4.0.3 (2020-10-10)
#>  os       Windows 10 x64              
#>  system   x86_64, mingw32             
#>  ui       RTerm                       
#>  language (EN)                        
#>  collate  English_United States.1252  
#>  ctype    English_United States.1252  
#>  tz       America/New_York            
#>  date     2021-02-03                  
#> 
#> - Packages -------------------------------------------------------------------
#>  package     * version date       lib source        
#>  assertthat    0.2.1   2019-03-21 [1] CRAN (R 4.0.0)
#>  bit           4.0.4   2020-08-04 [1] CRAN (R 4.0.2)
#>  bit64         4.0.5   2020-08-30 [1] CRAN (R 4.0.3)
#>  cli           2.3.0   2021-01-31 [1] CRAN (R 4.0.3)
#>  crayon        1.4.0   2021-01-30 [1] CRAN (R 4.0.3)
#>  digest        0.6.27  2020-10-24 [1] CRAN (R 4.0.3)
#>  ellipsis      0.3.1   2020-05-15 [1] CRAN (R 4.0.0)
#>  evaluate      0.14    2019-05-28 [1] CRAN (R 4.0.0)
#>  fs            1.5.0   2020-07-31 [1] CRAN (R 4.0.2)
#>  glue          1.4.2   2020-08-27 [1] CRAN (R 4.0.3)
#>  highr         0.8     2019-03-20 [1] CRAN (R 4.0.0)
#>  hms           1.0.0   2021-01-13 [1] CRAN (R 4.0.3)
#>  htmltools     0.5.1.1 2021-01-22 [1] CRAN (R 4.0.3)
#>  knitr         1.31    2021-01-27 [1] CRAN (R 4.0.3)
#>  lifecycle     0.2.0   2020-03-06 [1] CRAN (R 4.0.0)
#>  magrittr      2.0.1   2020-11-17 [1] CRAN (R 4.0.3)
#>  pillar        1.4.7   2020-11-20 [1] CRAN (R 4.0.3)
#>  pkgconfig     2.0.3   2019-09-22 [1] CRAN (R 4.0.0)
#>  ps            1.5.0   2020-12-05 [1] CRAN (R 4.0.3)
#>  purrr         0.3.4   2020-04-17 [1] CRAN (R 4.0.0)
#>  R6            2.5.0   2020-10-28 [1] CRAN (R 4.0.3)
#>  readr       * 1.4.0   2020-10-05 [1] CRAN (R 4.0.3)
#>  reprex        1.0.0   2021-01-27 [1] CRAN (R 4.0.3)
#>  rlang         0.4.10  2020-12-30 [1] CRAN (R 4.0.3)
#>  rmarkdown     2.6     2020-12-14 [1] CRAN (R 4.0.3)
#>  rstudioapi    0.13    2020-11-12 [1] CRAN (R 4.0.3)
#>  sessioninfo   1.1.1   2018-11-05 [1] CRAN (R 4.0.0)
#>  stringi       1.5.3   2020-09-09 [1] CRAN (R 4.0.3)
#>  stringr       1.4.0   2019-02-10 [1] CRAN (R 4.0.0)
#>  tibble        3.0.6   2021-01-29 [1] CRAN (R 4.0.3)
#>  tidyselect    1.1.0   2020-05-11 [1] CRAN (R 4.0.0)
#>  vctrs         0.3.6   2020-12-17 [1] CRAN (R 4.0.3)
#>  vroom       * 1.4.0   2021-02-01 [1] CRAN (R 4.0.3)
#>  withr         2.4.1   2021-01-26 [1] CRAN (R 4.0.3)
#>  xfun          0.20    2021-01-06 [1] CRAN (R 4.0.3)
#>  yaml          2.2.1   2020-02-01 [1] CRAN (R 4.0.0)
#> 
#> [1] C:/Program Files/R/Library/4.0
#> [2] C:/Program Files/R/R-4.0.3/library
```

</details>

**Comment by jimhester** [comment]
You need to load readr before vroom, this is just the way that S3 method registration works, there is little we can do about it unfortunately.

**Comment by kent37** [comment]
This also fails, it accesses `readr` before `vroom`.
``` r
xx <- readr::col_character()
df <- vroom::vroom(vroom::vroom_example("mtcars.csv"), col_select=-mpg)
vroom::spec(df)
#> NULL
```

I encountered this in a package that uses both readr and vroom. It always accesses them with package-qualified names, as above. This problem is a regression, I did not have any trouble before upgrading to vroom 1.4. Any thoughts on a work-around would be very welcome.

**Comment by kent37** [comment]
Also this:
``` r
library(readr)
library(vroom)
df <- vroom::vroom(vroom::vroom_example("mtcars.csv"), col_select=-mpg)
vroom::spec(df)
#> NULL
```
and this:
``` r
library(vroom)
library(readr)
#> Registered S3 methods overwritten by 'readr':
#>   method           from 
#>   format.col_spec  vroom
#>   print.col_spec   vroom
#>   print.collector  vroom
#>   print.date_names vroom
#>   print.locale     vroom
#>   str.col_spec     vroom
#> 
#> Attaching package: 'readr'
#> The following objects are masked from 'package:vroom':
#> 
#>     as.col_spec, col_character, col_date, col_datetime, col_double,
#>     col_factor, col_guess, col_integer, col_logical, col_number,
#>     col_skip, col_time, cols, cols_condense, cols_only, date_names,
#>     date_names_lang, date_names_langs, default_locale, fwf_cols,
#>     fwf_empty, fwf_positions, fwf_widths, locale, output_column,
#>     problems, spec
df <- vroom::vroom(vroom::vroom_example("mtcars.csv"), col_select=-mpg)
vroom::spec(df)
#> NULL
```

**Comment by jimhester** [comment]
Fixed by 69df3e44e84893

**Comment by kent37** [comment]
Thank you!
